ENH: `float.hex` and `float.fromhex` Implementation

ganesh-k13 opened the pull request:
### Adding `float.hex` and `float.fromhex` Implementation
Added `float.hex` as follows:
```py
>>> np.float32(3074).hex()
'0x1.804000p+11'
```

Added `float.fromhex`  as follows
```py
>>> np.float32.fromhex('0x1.804000p+11')
3074.0
```

### TODO

- [x] Documentation
- [x] Annotations [Already present for 64]
- [x] Testcases
    - [x] test_ends
    - [x] test_invalid_inputs
    - [x] test_whitespace
    - [x] test_from_hex
    - [x] test_roundtrip
    - [ ] ~test_subclass ?~
- [x] Release notes
- [ ] Benchmarks

Part of #13375

### References:
- https://math.stackexchange.com/questions/4265905/why-increase-the-mantissa-while-converting-float-to-hex-in-python
- https://github.com/python/cpython/commit/65fe25e59710eada7eb41bddb45bdca8b61b5668#diff-3651445c9f72daa1fdf6577b10902d4ceee89aa99295e01fcab07b58767a935fR1105

ganesh-k13 committed:
ENH: Added float.hex implementation

ganesh-k13 committed:
ENH: Added float.fromhex implementation

ganesh-k13 commented:
It was just easier to add `fromhex` than implement `hex` and `fromhex` separately for testing, etc.

ganesh-k13 commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -900,3 +902,398 @@ npy_popcount@c@(npy_@type@ a)
+    if (s != s_end)
+        goto parse_error;
+    *result = negate ? -x : x;
+    // if (type != &PyFloat_Type && result != NULL) {
```
Will remove this, forgot while changing the function

ganesh-k13 committed:
TST: Added basic `hex` and `fromhex` tests

ganesh-k13 committed:
ENH: Various overflow fixes

ganesh-k13 commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -2019,6 +2020,60 @@ static PyObject *
+        return NULL;
+    }
+
+    ret = PyArrayScalar_New(@Name@);
+    if (ret == NULL) {
+        return NULL;
+    }
+    PyArrayScalar_ASSIGN(ret, @Name@, value);
```
Hey @seberg / @mattip / @eric-wieser , apologies for the ping, I could not figure this out. This logic is not working for `flaot16` for some reason. For input say `>>> np.float16.fromhex(np.float16(2.0).hex())`, this is the trace:
```c
2073        PyArrayScalar_ASSIGN(ret, @Name@, value);
(gdb) 
2075        return ret;
(gdb) p value
$1 = 2
(gdb) c
Continuing.
1e-07
```
It seems like result is `1e-07` although the logic gives result(`value`) as `2`. Any pointers here will be very helpful. Thanks.

seberg replied:
The half path uses double (a bit weird that not float?), and you need to convert float/double to half first to store it, since there is no C `half` type.

ganesh-k13 replied:
Ohh I see. Shall I use float? I had the same doubt but referred to here where double was used and wondered why: https://github.com/numpy/numpy/blob/c8fdb53ee2d1b1130a74ad7a3250366fd042da3f/numpy/core/src/multiarray/scalartypes.c.src#L2005
Thanks a lot for the reply!

seberg replied:
Doesn't matter probably, if you are pretty sure there is no issue (and why would there be?) changing to float seems reasonable, since that is what most places do AFAIK.

ganesh-k13 replied:
It worked, thanks a lot!

ganesh-k13 committed:
ENH: Removed npy_half_* functions

ganesh-k13 commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -2019,6 +2020,80 @@ static PyObject *
+    // xref: https://stackoverflow.com/a/15197157/5671364
+    char buf[
+        1 + // sign, '-' or '+'
+#if @is_half@
+        (NPY_SIZEOF_FLOAT * CHAR_BIT + 3) / 4 + // mantissa hex digits max
+#else
+        (NPY_SIZEOF_@NAME@ * CHAR_BIT + 3) / 4 + // mantissa hex digits max
+#endif
+        1 + // decimal point, '.'
+        1 + // mantissa-exponent separator, 'p'
+        1 + // mantissa sign, '-' or '+'
+#if @is_half@
+        (NPY_SIZEOF_FLOAT * CHAR_BIT + 2) / 3 + // exponent decimal digits max
+#else
+        (NPY_SIZEOF_@NAME@ * CHAR_BIT + 2) / 3 + // exponent decimal digits max
+#endif
```
There is a staggering amount of `#if @is_half@` :), open to any suggestions to avoid them. We can convert the `PyObject` of `half` itself to `float` and try perhaps?

ganesh-k13 replied:
Kinda resolved in 07b4614, not very happy with the solution, hence keeping this thread open to suggestions,

ganesh-k13 committed:
TST: Added invalid inputs cases

ganesh-k13 commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -900,3 +902,390 @@ npy_popcount@c@(npy_@type@ a)
+    s = string;
+    if (s == NULL)
+        return NPY_FALSE;
+    s_end = s + strlen(s);
```
This logic is incorrect as we can inject a `\0` as shown [here](https://github.com/numpy/numpy/pull/20083/files#diff-c5ebcf9e35c56cb23da228c4bea6471c8587296ea82c5a0ee3da5fe3f6dbc741R287).

I'll try to compute length in a different way or worst case, pass it an arg to this function or maybe pass the start and end of the string in an iterator style.

ganesh-k13 replied:
Handled in 2ba28ca

ganesh-k13 committed:
ENH: Handle NULL in input string

ganesh-k13 committed:
TST: Added Whitespace tests

ganesh-k13 committed:
TST: Added test_from_hex

Added the following tests:
* two spellings of infinity, with optional signs; case-insensitive
* nans with optional sign; case insensitive
* variations in input format
* moving the point around
* [WIP] results that should overflow...
* ...and those that round to +-max float
* zeros

ganesh-k13 commented on numpy/core/tests/test_scalar_methods.py:
```diff
@@ -201,3 +202,227 @@ def test_bit_count(self):
+
+        # results that should overflow...
+        large_values = [
+            f"-0x1p{MAXEXP}", f"0x1p+{MAXEXP + 1}", f"+0X1p{MAXEXP + 5}",
```
Due to the fact we use `float32.fromhex` for `16`, the return value on underflow does not set an error when it overflows for `16` and simply leads to `(-)inf`. Need to see how we can handle this properly instead of checking at the caller again.

ganesh-k13 replied:
resolved in 07b4614

ganesh-k13 commented on numpy/core/tests/test_scalar_methods.py:
```diff
@@ -201,3 +202,227 @@ def test_bit_count(self):
+            # for 128. It can also be `9` for 64, etc
+            f"-0X1.{'f' * (NMANT // 4)}fp{MAXEXP - 1}",
+            f"+0x3.{'f' * (NMANT // 4 + 1)}p{MAXEXP - 2}",
+            # XXX How are we getting to 970?
+            # f"0x3fffffffffffffp+{MAXEXP - NMANT}",
```
Hi @mdickinson, can you please help me understand how we arrive at `970` for `64` bits? It exactly gives `1.7976931348623158079e+308` when max is `1.7976931348623157e+308`. Please ignore `{MAXEXP - NMANT}`, it was just a placeholder.

This is in reference to https://github.com/python/cpython/blob/e18d81569fa0564f3bc7bcfd2fce26ec91ba0a6e/Lib/test/test_float.py#L1332

Also, thanks for the extensive testsuite, was able to find a myriad of bugs :).

mdickinson replied:
`0x3fffffffffffffp+970` represents the smallest positive value that overflows rather than rounding to a finite number.

As to derivation: the maximum finite representable value is `2**DBL_MAX_EXP - 2**(DBL_MAX_EXP - DBL_MANT_DIG)` (in terms of the constants defined in the C standard). However, this isn't the boundary at which we start to overflow: that boundary is halfway between the above value and `2**DBL_MAX_EXP`, so it's `2**DBL_MAX_EXP - 2**(DBL_MAX_EXP - DBL_MANT_DIG - 1)`, which assuming IEEE 754 binary64, is `2**1024 - 2**970`, or `(2**54 - 1) * 2**970`.

For IEEE 754 single-precision (float32), the corresponding boundary would be `2**128 - 2**103`, or `0x1ffffffp+103`. For `float16`, it would be `2**16 - 2**4` (`0xfffp+4`):

A bit off-topic, but there's something a bit funny going on with int-to-float32 conversions in NumPy. I suspect a double-rounding through float64 first.

```python
>>> np.float32(2**128 - 2**103)  # ok: gives infinity, as expected
inf
>>> np.float32(2**128 - 2**103 - 1)  # bad: expect a finite result, get infinity instead
inf
```

It looks as though the actual boundary is at `2**128 - 2**103 - 2**74` instead of `2**128 - 2**103`:

```
>>> np.float32(2**128 - 2**103 - 2**74)
inf
>>> np.float32(2**128 - 2**103 - 2**74 - 1)
3.4028235e+38
```

ganesh-k13 replied:
Thanks a lot for the explanation @mdickinson! Will incorporate the same.

ganesh-k13 commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -900,3 +902,385 @@ npy_popcount@c@(npy_@type@ a)
+/* Heavily copied from the builtin float.fromhex
+ * This function returns True if conversion is possible,
+ * else returns false and value of the argument `result`
+ * should not be used.
```
Not sure why I added this. Will remove.

ganesh-k13 replied:
resolved in 07b4614

mdickinson commented:
> A bit off-topic, but there's something a bit funny going on with int-to-float32 conversions in NumPy.

I opened #20687 for this.

ganesh-k13 committed:
ENH: Refined handling of half

Improved half handling via:
* Added a new utility function to convert float hex to half
* Split half `hex` and `fromhex` to avoid macro clutter
* Special handling of overflows for half cases
* Few minor comments and string changes.

ganesh-k13 commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -2018,6 +2049,119 @@ static PyObject *
+        return NULL;
+    }
+
+    npy_clear_floatstatus_barrier((char *)&value);
+    half_value = npy_float_to_half(value);
+    if (npy_get_floatstatus_barrier((char *)&value) & NPY_FPE_OVERFLOW) {
```
I just realized after pushing, if we go beyond float instead of beyond half, we'll return at line 2094's `if` and the message will be `.. too large for float` instead of `half`. Hmm, half is getting real tricky with a lot of edge cases. Will see how we can handle this.

mattip replied:
Did this get handled?

seberg replied:
I think wouldn't worry about this one.  Yes, it would be nicer to report half, but overall it doesn't matter too much.

Rather than float barrier, it may be a bit faster/easier to explicitly check for an `inf` result (and that it was not `inf` before) for scalar cases.

seberg replied:
That change probably fixes the piodide error indirectly:  It doesn't properly support FPEs.

ganesh-k13 replied:
As a compromise, I removed the `@name@` so we don't get an incorrect message.

ganesh-k13 committed:
TST: Added test_from_hex and test_ends

test_from_hex:
* values that should underflow to 0

test_ends:
* Added case for MAX, MIN, TINY(smallest subnormal) and EPS

Misc:
* Added asserts. No idea what I was testing without it :)
* Various linting fixes

BvB93 reviewed (changes requested):
Can you update the `fromhex` signature in the .pyi stub file? Since currently it works with `np.float64` types.
``` diff
--- a/numpy/__init__.pyi
+++ b/numpy/__init__.pyi
@@ -2910,7 +2910,7 @@ def tolist(self) -> float: ...
     @classmethod
-    def fromhex(cls: type[float64], string: str, /) -> float64: ...
+    def fromhex(cls: type[_FloatType], string: str, /) -> _FloatType: ...
```

ganesh-k13 commented:
Thanks @BvB93, I was wondering the same. Will add to TODO

ganesh-k13 committed:
TST: Added test_from_hex

Added following cases:
* check round-half-even is working correctly near 0
* b.p.o. 44954 cases

Added TODO for following:
* Overflow end cases
* those that round to +-max float
* ... and near MIN
* ... and near 1.0
* Use proper Mantissa for moving the point around

ganesh-k13 committed:
ENH: Added annotations for fromhex

ganesh-k13 commented on numpy/__init__.pyi:
```diff
@@ -2908,9 +2908,9 @@ class floating(inexact[_NBit1]):
     def is_integer(self) -> bool: ...
-    def hex(self: float64) -> str: ...
+    def hex(self: _FloatType) -> str: ...
     @classmethod
-    def fromhex(cls: type[float64], string: str, /) -> float64: ...
+    def fromhex(cls: type[_FloatType], string: str, /) -> _FloatType: ...
```
@BvB93 , I added for `hex` and `fromhex` as I added both functions. This is correct?

ganesh-k13 committed:
DOC: Added docs for `hex` and `fromhex`

Added TODO for 128 bit precision loss

BvB93 commented on numpy/__init__.pyi:
```diff
@@ -2908,9 +2908,9 @@ class floating(inexact[_NBit1]):
     def tolist(self) -> float: ...
     def is_integer(self) -> bool: ...
-    def hex(self: float64) -> str: ...
+    def hex(self: _FloatType) -> str: ...
```
```suggestion
    def hex(self) -> str: ...
```
No need to use a typevar here, as you're dealing with <2 variables here.
`fromhex` looks fine though.

ganesh-k13 replied:
Thanks, fixed in f48f0fc

ganesh-k13 committed:
DOC: `hex` and `fromhex` release notes (#20083)

BvB93 commented on doc/release/upcoming_changes/20083.new_feature.rst:
```diff
@@ -0,0 +1,16 @@
+`float.hex <https://docs.python.org/3/library/stdtypes.html#float.hex>`__ and
+`float.fromhex <https://docs.python.org/3/library/stdtypes.html#float.fromhex>`__
+
+.. code-block:: python
```
Could be mistaken, but I don't think `.. code-block` renders correctly without an empty line.
```suggestion
.. code-block:: python

```

ganesh-k13 replied:
Ah my bad, will fix this. This unfortunately does not render in docs build making it a bit hard.

ganesh-k13 replied:
Fixed in 34593e7

ganesh-k13 commented:
Hi team, I'm opening this for review. Now there are two major things remaining which I'll pick up in a following PR:
1. Testing items for `fromhex`, I have added appropriate `TODO`s and will open issues for same once this is approved/merged. (I realised how little of floats I know in the process, need to read up to complete this)
2. `Half` overflow errors will look weird, courtesy: https://github.com/numpy/numpy/pull/20083#discussion_r777116965

ganesh-k13 commented:
Hey @seberg , @mattip, @BvB93  adding you as reviewers so it's in your backlog. Please have a look when free. Thanks!

seberg commented:
We discussed this and should push this forward.  At this point review is the most important part probably.  I will try to get to it soon (other reviews of course great :)).

seberg commented:
I have to admit, this is more complicated than I expected.  Especially about long-double (since we ideally should not hardcode the number of mantissa bits for rounding).

Is there a big reason to avoid the C99 `strtod` and `strtold` functions?  They are C99, so we presumably should be able to rely on them?
Yes, it requires that locale dance (and if a system fails to provide `strtold`, we will have to bail), but is that bad enough to ship a custom version?
It also seems `fromhex` is stricter about overflows than most float parsing.  But it seems to me that can be handled by postprocessing (checking whether the result is inf).

Or is there a very important reason that Python ships its own version, beyond not wanting to rely on on C99 (at the time)?

mdickinson commented:
> Or is there a very important reason that Python ships its own version, beyond not wanting to rely on on C99 (at the time)?

Not wanting to rely on C99 would have been the main reason (in particular, it seems likely that MSVC's `strtod` wouldn't have supported hex floating-point at the time). But separately from that, it seemed desirable to have correct rounding, and I wouldn't have trusted library implementations of `strtod` to round correctly in all cases. (I still don't: there's nothing in the C standard that requires it, and I wouldn't be at all surprised to find that some implementations simply truncate the input before processing.)

Faithful rounding is obviously essential here, but maybe correct rounding is a nice-to-have that NumPy doesn't need?

mdickinson commented:
> there's nothing in the C standard that requires it

I take it back. It's right there in C99 §7.20.1.3p5:

> If the subject sequence has the hexadecimal form and FLT_RADIX is a power of 2, the value resulting from the conversion is correctly rounded.

I'm still not sure I'd trust platform libs to actually do that correct rounding, though.

Slightly off-topic, but it also occurs to me that these days, CPython should also be accepting underscores in the input to `fromhex` (e.g., allowing `float.fromhex("0x1.921f_b544_42d1_8p+1")`). But that's off-topic for this tracker.

seberg commented:
Well, missing MSVC support is a problem :).  I also won't trust correct rounding, at least not for longdouble.  My problem is right now I am not sure I trust this code for long double rounding, either :).
An alternative might be to steal the double version from Python (also adapting to float) and use the system one for long double (if available, if not, we fall back to the double version.  If long double size is double we do that always, so MSVC is out of the equation in any case).

If @ganesh-k13 is confident about the subtleties of long double, then I am OK to go with this also (right now, I am simply unsure).  I also half expect we could reformulate things to not need to know how many bits the mantissa has (ahead of time), but solving that puzzle might just be a waste of time if this or the C99 functions are good :).

ganesh-k13 committed:
TST: Added 64 bit specific tests

Tracking in https://github.com/numpy/numpy/issues/21903

1. moving the point around (pi).
2. Make it go towards NMANT (in ref to [1])
3. results that should overflow...
4. check round-half-even is working correctly near MIN
5. check round-half-even is working correctly near 1.0

ganesh-k13 commented:
> An alternative might be to steal the double version from Python (also adapting to float) and use the system one for long double (if available, if not, we fall back to the double version

This seems like a viable option, shall we do it right away or let this one go in and add conditions like we did for `popcount` https://github.com/numpy/numpy/blob/3d9cacc100e1b9382a543f000d0df44180cff122/numpy/core/src/npymath/npy_math_internal.h.src#L887

> If @ganesh-k13 is confident about the subtleties of long double, then I am OK to go with this also (right now, I am simply unsure).

If I'm being honest, I wasn't sure if what I ported from Cpython works for anything other that 64 bits, but the tests were pretty exptensive and we cover the important edgecases for all sizes. So I guess we are ok with this?

> CPython should also be accepting underscores in the input to fromhex
Ah ok, will add it to my backlog. Thanks for the heads up

ganesh-k13 commented:
Hey folks, pinging again on this. I think the long double part is OK, I don't see anything breaking as the tests are pretty comprehensive (taken from CPython). Anything I can do to push this through?

mattip commented:
there is a merge conflict, and this also needs another round of review

ganesh-k13 commented:
I can resolve the merge conflict this week. @mattip any idea who can review this? Can Mark from CPython help? I can try pinging him.

mattip commented:
Let's see once it is ready for review.

ganesh-k13 commented:
i have resolved the conflicts

ganesh-k13 commented:
@mattip any pointers on how to proceed with this one?

mattip commented:
There are some failing tests. Is there a problem with Pyodide and PyPy?

Could you add a test that the result of `f.hex()` is equivalent in NumPy and CPython, for some values of f (as a python float vs a NumPy scalar)

seberg reviewed (commented):
Not a full review, but some comments (not all need addressing).

Overall, I guess we should just go with this.

I think right now these functions end up as public API in `npy_math`.  That doesn't seem bad (they are very well defined), just wanted to check on it briefly.

ganesh-k13 committed:
MAINT: Changed `fromhex` return type

ganesh-k13 committed:
MAINT: Changed half overflow detection

ganesh-k13 commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -915,3 +917,385 @@ npy_popcount@c@(npy_@type@ a)
+ * should be used as resultant floating point value.
+ */
+NPY_INPLACE npy_bool
+npy_@name@_fromhex(const char *s, const char *s_end, npy_@name@ *result)
```
Edited in [8d7bc55](https://github.com/numpy/numpy/pull/20083/commits/8d7bc55d26ec5b3db35270a4f13b45ae68cafeb3)

ganesh-k13 commented on numpy/core/_add_newdocs_scalars.py:
```diff
@@ -291,6 +291,42 @@ def add_newdoc_for_scalar_type(obj, fixed_aliases, doc):
+        ('longdouble', ('-0x1.999999999999a000p-4', '0x1.921fb54442d18000p+1'))
+    ):
+    # TODO: We loose precision for 128 bits here, hence the ending 0s
+    pi_rounded = getattr(_numerictypes, float_name)(3.141592653589793116)
```
Oh, I did not fully get it, Do you mean change it to individual type's string so it's uniform across platforms?

ganesh-k13 commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -1991,9 +1991,40 @@ error:
+    // read mantissa (0xh.hh) and store end at mant_end
+    // Size of mantissa + decimal '.'
+    mant_end = i + (NPY_SIZEOF_HALF * CHAR_BIT + 3) / 4 + 1 + 1;
+    i = mant_end = mant_end+1;
```
I think we don't need to round here. I don't have a mathematical proof as to why we don't need to round off, however, purely guessing from the tests passing.

ganesh-k13 commented:
The explicit `inf` check helped with Pyodide issue. I'm curious how the other calls are passing. Is it a lack of tests or is the caller handling it?
